File: api/app/infrastructure/external/sandbox/docker_sandbox.py
from typing import BinaryIO, Self
from fastapi import UploadFile
import logging
from app.domain.models.tool_result import ToolResult
from app.infrastructure.external.browser.playwright_browser import PlayWrightBrowser
from app.domain.external.browser import Browser
import uuid
import asyncio
from core.config import get_settings
import httpx
import socket
from typing import Optional
from app.domain.external.sandbox import Sandbox
from async_lru import alru_cache
import docker
import io

logger = logging.getLogger(__name__)


class DockerSandbox(Sandbox):
    def __init__(self, ip: Optional[str] = None, container_name: Optional[str] = None):
        self.client = httpx.AsyncClient(timeout=600)
        self._ip = ip
        self._container_name = container_name
        self._base_url = f"http://{self._ip}:8000"
        self._cdp_url = f"http://{self._ip}:9222"
        self._vnc_url = f"http://{self._ip}:5900"

    @property
    def id(self) -> str:
        if self._container_name is None:
            return "manus-sandbox"
        return self._container_name

    @property
    def cdp_url(self) -> str:
        return self._cdp_url

    @property
    def vnc_url(self) -> str:
        return self._vnc_url

    @classmethod
    @alru_cache(maxsize=128, typed=True)
    async def _resolve_hostname_to_ip(cls, hostname: str) -> str:
        try:
            try:
                socket.inet_pton(socket.AF_INET, hostname)
                return hostname
            except OSError:
                pass

            addr_info = socket.getaddrinfo(hostname, None, family=socket.AF_INET)
            if addr_info and len(addr_info) > 0:
                return addr_info[0][4][0]
            else:
                return None
        except Exception as e:
            raise ValueError(f"Failed to resolve hostname {hostname}: {e}")
            return None

    @classmethod
    def _get_container_ip(cls, container: docker.models.containers.Container) -> str | None:
        network_settings = container.attrs.get("NetworkSettings", {})
        ip = network_settings.get("IPAddress")
        if not ip:
            for network in network_settings.get("Networks", {}).values():
                ip = network.get("IPAddress")
                if ip:
                    break
        return ip

    @classmethod
    def _create_task(cls) -> Self:
        settings = get_settings()

        image = settings.sandbox_image
        name_prefix = settings.sandbox_name_prefix
        container_name = f"{name_prefix}-{uuid.uuid4()}"

        try:
            client = docker.from_env()
            container_config = {
                "image": image,
                "name": container_name,
                "detach": True,
                "remove": True,
                "ports": {
                    "8000/tcp": 8000,
                    "9222/tcp": 9222,
                    "5900/tcp": 5900,
                    "5901/tcp": 5901,
                },
                "environment": {
                    "SERVICE_TIMEOUT_MINUTES": settings.sandbox_ttl_minutes,
                    "CHROME_ARGS": settings.sandbox_chrome_args,
                    "HTTPS_PROXY": settings.sandbox_https_proxy,
                    "HTTP_PROXY": settings.sandbox_http_proxy,
                    "NO_PROXY": settings.sandbox_no_proxy,
                },
                "volumes": {
                    "/dev/shm": {
                        "bind": "/dev/shm",
                        "mode": "rw",
                    }
                },
            }

            if settings.sandbox_network:
                container_config["network"] = settings.sandbox_network

            container = client.containers.run(**container_config)
            container.reload()
            logger.info(f"Sandbox container {container_name} created")
            ip = cls._get_container_ip(container)
            logger.info(f"Sandbox container {container_name} IP: {ip}")
            return cls(ip="127.0.0.1", container_name=container_name)
        except Exception as e:
            logger.error(f"Failed to create sandbox: {e}")
            raise Exception(f"Failed to create sandbox: {e}")

    @classmethod
    async def create(cls) -> Self:
        settings = get_settings()

        if settings.sandbox_address:
            ip = await cls._resolve_hostname_to_ip(settings.sandbox_address)
            return cls(ip="127.0.0.1", container_name=settings.sandbox_name_prefix)

        return await asyncio.to_thread(cls._create_task)

    async def destroy(self) -> bool:
        try:
            if self.client:
                await self.client.aclose()

            if self._container_name:
                container = docker.from_env().containers.get(self._container_name)
                container.stop(timeout=5)
                container.remove(force=True)
            return True
        except Exception as e:
            raise Exception(f"Failed to destroy sandbox: {e}")

    @classmethod
    async def get(cls, id: str) -> Optional[Self]:
        settings = get_settings()
        if settings.sandbox_address:
            try:
                ip = await cls._resolve_hostname_to_ip(settings.sandbox_address)
                return cls(ip=ip, container_name=settings.sandbox_name_prefix)
            except Exception as e:
                logger.error(f"Failed to get sandbox: {e}")
                return None

        try:
            docker_client = docker.from_env()
            try:
                container = docker_client.containers.get(id)
                container.reload()

                if container.status != "running":
                    logger.error(f"Sandbox container {id} is not running")
                    return None

                ip = cls._get_container_ip(container)

                if not ip:
                    logger.error(f"Sandbox container {id} has no IP address")
                    return None

                return cls(ip='127.0.0.1', container_name=id)
            except docker.errors.NotFound:
                logger.error(f"Sandbox container {id} not found")
                return None
            except docker.errors.APIError as e:
                logger.error(f"Failed to get sandbox: {e}")
                return None
            finally:
                docker_client.close()
        except Exception as e:
            logger.error(f"Failed to get sandbox: {e}")
            return None

    async def get_browser(self) -> Browser:
        return PlayWrightBrowser(self.cdp_url)

    async def _ensure_sandbox_exists(self) -> None:
        max_retries = 30
        retry_interval = 2
        for attempt in range(max_retries):
            try:
                response = await self.client.get(
                    f"{self._base_url}/api/supervisor/status"
                )
                response.raise_for_status()
                tool_result = ToolResult.from_sandbox(**response.json())

                if not tool_result.success:
                    logger.warning(
                        f"无法确认sandbox Supervisor状态, {str(tool_result.message)}"
                    )
                    await asyncio.sleep(retry_interval)
                    continue

                services = tool_result.data or []

                if not services:
                    logger.warning("sandbox Supervisor进程中未发现任何服务")
                    await asyncio.sleep(retry_interval)
                    continue

                all_running = True
                non_runnning_services = []

                for service in services:
                    if service.get("statename") != "RUNNING":
                        all_running = False
                        non_runnning_services.append(service.get("name"))

                if not all_running:
                    logger.warning(
                        f"sandbox Supervisor进程中存在未运行的服务: {non_runnning_services}"
                    )
                    await asyncio.sleep(retry_interval)
                    continue

                return
            except Exception as e:
                if attempt == max_retries - 1:
                    raise Exception(f"Failed to ensure sandbox: {e}")
                await asyncio.sleep(retry_interval)

    # file模块方法
    async def file_read(
        self,
        filepath: str,
        start_line: Optional[int] = None,
        end_line: Optional[int] = None,
        sudo: Optional[bool] = False,
    ) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/file/read-file",
            json={
                "filepath": filepath,
                "start_line": start_line,
                "end_line": end_line,
                "sudo": sudo,
            },
        )
        response.raise_for_status()
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

    async def file_write(
        self,
        filepath: str,
        content: str,
        append: bool = False,
        leading_newline: bool = False,
        trailing_newline: bool = False,
        sudo: bool = False,
    ) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/file/write-file",
            json={
                "filepath": filepath,
                "content": content,
                "append": append,
                "leading_newline": leading_newline,
                "trailing_newline": trailing_newline,
                "sudo": sudo,
            },
        )
        response.raise_for_status()
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

    async def file_replace(
        self,
        filepath: str,
        old_str: str,
        new_str: str,
        sudo: bool = False,
    ) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/file/replace-in-file",
            json={
                "filepath": filepath,
                "old_content": old_str,
                "new_content": new_str,
                "sudo": sudo,
            },
        )
        response.raise_for_status()
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

    async def file_search(
        self,
        filepath: str,
        regex: str,
        sudo: bool = False,
    ) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/file/search-in-file",
            json={
                "filepath": filepath,
                "regex": regex,
                "sudo": sudo,
            },
        )
        response.raise_for_status()
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

    async def file_find(self, dirpath: str, glob_pattern: str) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/file/find-files",
            json={
                "dir_path": dirpath,
                "glob_pattern": glob_pattern,
            },
        )
        response.raise_for_status()
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

    async def file_list(self, dirpath: str) -> ToolResult:
        return await self.file_find(dirpath, "*")

    async def file_upload(
        self, file_data: BinaryIO, filepath: str, filename: str
    ) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/file/upload-file",
            json={
                "file": file_data,
                "filepath": filepath,
                "filename": filename,
            },
        )
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

    async def check_file_exists(self, filepath: str) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/supervisor/check-file-exists",
            json={
                "filepath": filepath,
            },
        )
        response.raise_for_status()
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

    async def file_delete(self, filepath: str, sudo: bool = False) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/supervisor/delete-file",
            json={
                "filepath": filepath,
                "sudo": sudo,
            },
        )
        response.raise_for_status()
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

    async def file_download(self, filepath: str, sudo: bool = False) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/supervisor/download-file",
            json={
                "filepath": filepath,
                "sudo": sudo,
            },
        )
        response.raise_for_status()
        return io.BytesIO(response.content)

    # shell模块
    async def exec_command(
        self, session_id: str, exec_dir: str, command: str
    ) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/shell/exec-command",
            json={
                "session_id": session_id,
                "command": command,
                "exec_dir": exec_dir,
            },
        )
        response.raise_for_status()
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

    async def view_shell(self, session_id: str, console: bool = False) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/shell/view_shell",
            json={
                "session_id": session_id,
                "console": console,
            },
        )
        response.raise_for_status()
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

    async def wait_for_process(
        self, session_id: str, seconds: Optional[str] = None
    ) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/shell/wait_for_process",
            json={
                "session_id": session_id,
                "seconds": seconds,
            },
        )
        response.raise_for_status()
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

    async def write_to_process(
        self, session_id: str, input_text: str, press_enter: bool = False
    ) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/shell/write-to-process",
            json={
                "session_id": session_id,
                "inputText": input_text,
                "enter": press_enter,
            },
        )
        response.raise_for_status()
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

    async def kill_process(self, session_id: str) -> ToolResult:
        response = await self.client.post(
            f"{self._base_url}/api/shell/kill-process",
            json={
                "session_id": session_id,
            },
        )
        response.raise_for_status()
        tool_result = ToolResult.from_sandbox(**response.json())
        return tool_result

File: api-node/src/infrastructure/sandbox/docker-sandbox.ts
import axios, { AxiosInstance } from 'axios';
import Docker from 'dockerode';
import fs from 'fs';
import dns from 'dns/promises';
import { v4 as uuidv4 } from 'uuid';
import { getSettings } from '../../config';
import { logger } from '../../utils/logger';

export interface ToolResult<T = unknown> {
  success: boolean;
  message?: string;
  data?: T;
  error?: string;
}

function fromSandboxResponse<T>(body: { code?: number; message?: string; data?: T }): ToolResult<T> {
  const success = body.code === 200;
  return {
    success,
    message: body.message,
    data: body.data,
    error: success ? undefined : body.message,
  };
}

function isRunningInDocker(): boolean {
  try {
    return fs.existsSync('/.dockerenv');
  } catch {
    return false;
  }
}

function createDockerClient(): Docker {
  const dockerHost = process.env.DOCKER_HOST;
  if (dockerHost) {
    if (dockerHost.startsWith('unix://')) {
      return new Docker({ socketPath: dockerHost.slice('unix://'.length) });
    }
    if (dockerHost.startsWith('npipe://')) {
      return new Docker({ socketPath: dockerHost.slice('npipe://'.length) });
    }
    return new Docker({ host: dockerHost });
  }
  if (process.platform === 'win32') {
    return new Docker({ socketPath: '//./pipe/docker_engine' });
  }
  return new Docker({ socketPath: '/var/run/docker.sock' });
}

function getDockerSocketPath(): string {
  const dockerHost = process.env.DOCKER_HOST;
  if (dockerHost?.startsWith('unix://')) return dockerHost.slice('unix://'.length);
  if (dockerHost?.startsWith('npipe://')) return dockerHost.slice('npipe://'.length);
  if (process.platform === 'win32') return '//./pipe/docker_engine';
  return '/var/run/docker.sock';
}

function wrapDockerError(err: unknown): Error {
  const message = err instanceof Error ? err.message : String(err);
  const code = err instanceof Error && 'code' in err ? String((err as NodeJS.ErrnoException).code) : '';

  if (code === 'ENOENT' && message.includes('docker.sock')) {
    return new Error(
      '无法连接 Docker 守护进程。动态创建沙箱需要 Docker 可用：\n'
      + '- 容器内运行 api-node 时挂载：-v /var/run/docker.sock:/var/run/docker.sock\n'
      + '- 本地 npm run dev 时确保 Docker Desktop 已启动',
    );
  }
  if (message.includes('No such image') || message.includes('pull access denied')) {
    return new Error(
      `沙箱镜像不存在或无法拉取，请先构建：docker build -t <SANDBOX_IMAGE> ./sandbox-node`,
    );
  }
  if (err instanceof Error) return err;
  return new Error(message);
}

function assertDynamicCreateReady(): void {
  const settings = getSettings();
  const socketPath = getDockerSocketPath();
  if (!fs.existsSync(socketPath)) {
    throw wrapDockerError(Object.assign(new Error(`connect ENOENT ${socketPath}`), { code: 'ENOENT' }));
  }
  if (isRunningInDocker() && !settings.sandboxNetwork) {
    throw new Error(
      '容器内动态创建沙箱必须设置 SANDBOX_NETWORK（与 api-node 相同，如 manus-network-dev），'
      + '否则新建的沙箱无法被 api-node 访问。',
    );
  }
}

export class DockerSandbox {
  private client: AxiosInstance;
  private readonly containerName: string | null;

  constructor(
    private readonly ip: string,
    containerName: string | null = null,
  ) {
    this.containerName = containerName;
    this.client = axios.create({
      baseURL: `http://${ip}:8000`,
      timeout: 600_000,
    });
  }

  get id(): string {
    return this.containerName ?? 'manus-sandbox';
  }

  get cdpUrl(): string {
    return `http://${this.ip}:9222`;
  }

  get vncUrl(): string {
    return `ws://${this.ip}:5901`;
  }

  private static getContainerIp(container: {
    NetworkSettings?: { IPAddress?: string; Networks?: Record<string, { IPAddress?: string }> };
  }): string | null {
    const networks = container.NetworkSettings?.Networks ?? {};
    for (const net of Object.values(networks)) {
      if (net?.IPAddress) return net.IPAddress;
    }
    return container.NetworkSettings?.IPAddress || null;
  }

  private static async resolveHostname(hostname: string): Promise<string> {
    try {
      return (await dns.lookup(hostname, { family: 4 })).address;
    } catch {
      return hostname;
    }
  }

  private static createContainer(): Promise<DockerSandbox> {
    assertDynamicCreateReady();

    const settings = getSettings();
    const docker = createDockerClient();
    const containerName = `${settings.sandboxNamePrefix}-${uuidv4()}`;
    const image = settings.sandboxImage ?? 'manus-sandbox-node';

    const hostConfig: Record<string, unknown> = {
      AutoRemove: true,
      Binds: ['/dev/shm:/dev/shm'],
    };
    if (settings.sandboxNetwork) {
      hostConfig.NetworkMode = settings.sandboxNetwork;
    }

    const config: Record<string, unknown> = {
      Image: image,
      name: containerName,
      HostConfig: hostConfig,
      Env: [
        `SERVICE_TIMEOUT_MINUTES=${settings.sandboxTtlMinutes}`,
        `CHROME_ARGS=${settings.sandboxChromeArgs ?? ''}`,
        `HTTPS_PROXY=${settings.sandboxHttpsProxy ?? ''}`,
        `HTTP_PROXY=${settings.sandboxHttpProxy ?? ''}`,
        `NO_PROXY=${settings.sandboxNoProxy ?? ''}`,
      ],
    };

    return new Promise((resolve, reject) => {
      docker
        .createContainer(config)
        .then((container) =>
          container
            .start()
            .then(() => container.inspect())
            .then((info) => {
              const ip = DockerSandbox.getContainerIp(info);
              if (!ip) {
                return reject(new Error(`Sandbox container ${containerName} has no IP`));
              }
              logger.info(`Sandbox container ${containerName} created at ${ip}`);
              resolve(new DockerSandbox(ip, containerName));
            }),
        )
        .catch((err) => reject(wrapDockerError(err)));
    });
  }

  static async create(): Promise<DockerSandbox> {
    const settings = getSettings();
    if (settings.sandboxAddress) {
      const ip = await DockerSandbox.resolveHostname(settings.sandboxAddress);
      return new DockerSandbox(ip, settings.sandboxNamePrefix);
    }
    return DockerSandbox.createContainer();
  }

  static async get(id: string): Promise<DockerSandbox | null> {
    const settings = getSettings();
    if (settings.sandboxAddress) {
      try {
        const ip = await DockerSandbox.resolveHostname(settings.sandboxAddress);
        return new DockerSandbox(ip, settings.sandboxNamePrefix);
      } catch (e) {
        logger.error(`Failed to get sandbox: ${e}`);
        return null;
      }
    }
    try {
      const docker = createDockerClient();
      const container = docker.getContainer(id);
      const info = await container.inspect();
      if (info.State?.Running !== true) {
        logger.error(`Sandbox container ${id} is not running`);
        return null;
      }
      const ip = DockerSandbox.getContainerIp(info);
      if (!ip) {
        logger.error(`Sandbox container ${id} has no IP`);
        return null;
      }
      return new DockerSandbox(ip, id);
    } catch (e) {
      logger.error(`Failed to get sandbox ${id}: ${e}`);
      return null;
    }
  }

  async destroy(): Promise<boolean> {
    if (!this.containerName) return true;
    try {
      const docker = createDockerClient();
      const container = docker.getContainer(this.containerName);
      await container.stop({ t: 5 });
      await container.remove({ force: true });
      return true;
    } catch (e) {
      throw new Error(`Failed to destroy sandbox: ${e}`);
    }
  }

  async ensureSandboxExists(): Promise<void> {
    const maxRetries = 30;
    const retryIntervalMs = 2000;

    for (let attempt = 0; attempt < maxRetries; attempt++) {
      try {
        const res = await this.client.get('/api/supervisor/status');
        const result = fromSandboxResponse(res.data);

        if (!result.success) {
          logger.warn(`Sandbox supervisor status check failed: ${result.message}`);
          await new Promise((r) => setTimeout(r, retryIntervalMs));
          continue;
        }

        const services = result.data;
        if (!Array.isArray(services) || services.length === 0) {
          logger.warn('Sandbox supervisor returned no services');
          await new Promise((r) => setTimeout(r, retryIntervalMs));
          continue;
        }

        const notRunning = services.filter((s: { statename?: string }) => s.statename !== 'RUNNING');
        if (notRunning.length > 0) {
          logger.warn(
            `Sandbox services not ready: ${notRunning.map((s: { name?: string }) => s.name).join(', ')}`,
          );
          await new Promise((r) => setTimeout(r, retryIntervalMs));
          continue;
        }

        return;
      } catch (e) {
        if (attempt === maxRetries - 1) {
          throw new Error(`Failed to ensure sandbox is ready: ${(e as Error).message}`);
        }
        await new Promise((r) => setTimeout(r, retryIntervalMs));
      }
    }
    throw new Error('Failed to ensure sandbox is ready');
  }

  async fileRead(filepath: string, startLine?: number, endLine?: number, sudo = false) {
    const res = await this.client.post('/api/file/read-file', {
      filepath,
      start_line: startLine,
      end_line: endLine,
      sudo,
    });
    return fromSandboxResponse(res.data);
  }

  async fileWrite(
    filepath: string,
    content: string,
    append = false,
    leadingNewline = false,
    trailingNewline = false,
    sudo = false,
  ) {
    const res = await this.client.post('/api/file/write-file', {
      filepath,
      content,
      append,
      leading_newline: leadingNewline,
      trailing_newline: trailingNewline,
      sudo,
    });
    return fromSandboxResponse(res.data);
  }

  async viewShell(sessionId: string, console = false) {
    const res = await this.client.post('/api/shell/view_shell', {
      session_id: sessionId,
      console,
    });
    return fromSandboxResponse(res.data);
  }

  async execCommand(sessionId: string, command: string, execDir: string) {
    const res = await this.client.post('/api/shell/exec-command', {
      session_id: sessionId,
      command,
      exec_dir: execDir,
    });
    return fromSandboxResponse(res.data);
  }

  async fileReplace(filepath: string, oldStr: string, newStr: string, sudo = false) {
    const res = await this.client.post('/api/file/replace-in-file', {
      filepath,
      old_content: oldStr,
      new_content: newStr,
      sudo,
    });
    return fromSandboxResponse(res.data);
  }

  async fileSearch(filepath: string, regex: string, sudo = false) {
    const res = await this.client.post('/api/file/search-in-file', {
      filepath,
      regex,
      sudo,
    });
    return fromSandboxResponse(res.data);
  }

  async fileFind(dirpath: string, globPattern: string) {
    const res = await this.client.post('/api/file/find-files', {
      dir_path: dirpath,
      glob: globPattern,
    });
    return fromSandboxResponse(res.data);
  }

  async fileList(dirpath: string) {
    return this.fileFind(dirpath, '*');
  }

  async fileUpload(fileData: Buffer, filepath: string, filename: string) {
    const form = new FormData();
    form.append('file', new Blob([fileData]), filename);
    form.append('filepath', filepath);
    const res = await this.client.post('/api/file/upload-file', form);
    return fromSandboxResponse(res.data);
  }

  async checkFileExists(filepath: string) {
    const res = await this.client.post('/api/file/check-file-exists', { filepath });
    return fromSandboxResponse(res.data);
  }

  async fileDelete(filepath: string, sudo = false) {
    const res = await this.client.post('/api/file/delete-file', { filepath, sudo });
    return fromSandboxResponse(res.data);
  }

  async fileDownload(filepath: string, sudo = false): Promise<Buffer> {
    const res = await this.client.get('/api/file/download-file', {
      params: { filepath },
      responseType: 'arraybuffer',
    });
    return Buffer.from(res.data);
  }

  async waitForProcess(sessionId: string, seconds?: string) {
    const res = await this.client.post('/api/shell/wait_for_process', {
      session_id: sessionId,
      seconds,
    });
    return fromSandboxResponse(res.data);
  }

  async writeToProcess(sessionId: string, inputText: string, pressEnter = false) {
    const res = await this.client.post('/api/shell/write-to-process', {
      session_id: sessionId,
      inputText,
      enter: pressEnter,
    });
    return fromSandboxResponse(res.data);
  }

  async killProcess(sessionId: string) {
    const res = await this.client.post('/api/shell/kill-process', { session_id: sessionId });
    return fromSandboxResponse(res.data);
  }
}
